### ISSUE TITLE
Certificate validation fails with unnecessary certificate in the CA chain

### ISSUE THREAD

**Issue by vokac** [eu-emi/canl-java#116]
When server presents unnecessary certificates to the client and one of them is not part of the trusted chain then CaNL library still try to validate all presented certificates. It seems to me that certificates that are not reachable by chain of parent CAs should be ignored / not used for decision if presented certificate with its chain can be trusted. This leads to verification failure in case on of presented and unused certificate (that is not part of trusted chain) can't pass CaNL validation.

This issue is described with more details in related [issue filed with dCache](https://github.com/dCache/dcache/issues/6662).

**Comment by paulmillar** [comment]
These two issues are very similar, but they aren't (necessarily) the same.  In #105, CaNL is the server and is validating client authentication.  In this ticket, it's the other way around: CaNL is the client, validating the certificates presented by the remote server.

IIRC, the validation framework in Java distinguishes between these two use-cases.  They *could* be handled identically within CaNL, but (AFAIK) this isn't guaranteed.

So, feel free to merge these two issues if you feel they are really the same thing.

**Comment by golbi** [comment]
@paulmillar I think that we have a common bug in both cases, and I'd be surprised if server/client side plays any role here. But thanks for heads up, will take that into account.

**Comment by vokac** [comment]
It seems to me that `atlas-amazon-cloud.cern.ch:443` still same configuration and advertise too much certificates (that are not part of trusted chain) while replying to the client that create TLS connection as described in the dCache [issue#6662](https://github.com/dCache/dcache/issues/6662). For us trust anchors are IGTF certificates and in this case it should be already intermediate `/DC=ch/DC=cern/CN=CERN Grid Certification Authority` from `ca_CERN-GridCA-1.116-1`.

**Comment by golbi** [comment]
@vokac @paulmillar Thx for the input. I've taken the 6 certificates as advertised by the atlas-amazon-... host, built a fresh IGTF truststore (1.116) and run two tests. 
* One was using simple canl DirectoryCertChainValidator, where trust anchors are just pem files from a directory. In the directory I've placed only the CERN grid CA cert.
* Another one was using OpensslCertChainValidator, where the complete IGTF truststore was configured.

In both cases validation succeeded. I've manually confirmed that the internal behaviour was as expected: only the first cert from chain was directly trusted by the discovered trust anchor. 

So seems that this problem is related to some particular way of canl lib is used by dCache (or other tool where the problem is observed). Paul, can you provide more details on how canl validator is setup in your case, and precisely how is invoked? Or any other problematic data set to run tests on?

**Comment by paulmillar** [comment]
Thanks for investigating further @golbi .

The SSLContext is initialised on pool start-up and cached for all subsequent HTTP client requests if the client doesn't supply a delegated X.509 (proxy) credential with the transfer.  If there's an X.509 credential then the SSLContext is built per-transfer (but cached for all requests related to that transfer).

Perhaps the most helpful entry point to see this is the [RemoteHttpTransferService](https://github.com/dCache/dcache/blob/master/modules/dcache/src/main/java/org/dcache/pool/classic/RemoteHttpTransferService.java) class.

The `createMoverProtocol` method is used to create the dCache-abstraction of data movement (the "MoverProtocol"), which handles the protocol-specific aspects of data transfer.

The method `buildSSLContext` builds a new (client) SSL context, when needed.

I hope the code is mostly self-explanatory, but just let me know if anything is unclear.

Cheers,
Paul.

**Comment by paulmillar** [comment]
Ping, @golbi  :smile: 

Do you need any more information in order to reproduce this problem?

Cheers,
Paul.

**Comment by golbi** [comment]
Hi @paulmillar 

I think I do. I have very limited time to invest into that lib, so a pull request with a failing testcase reproducing the problem would be very helpful. 

cheers, K

**Comment by paulmillar** [comment]
Hi @golbi 

I've created a pull-request that adds a (very modest) set of unit tests for the `OpensslCertChainValidator` class.  There is one test that is marked `@Ignore` that recreates the problem described in this issue.

(I've placed the code under an Apache 2.0 license.  I think this is compatible with CANL, but just let me know if a different license would be better.)

Cheers,
Paul.

**Comment by golbi** [comment]
Hi @paulmillar 

So what is going on. First of all your testcase has a serious bug :-) Order of certificates to be verified was reversed and it was starting with root CA. So the test was trying to validate root CA cert. 
It should be as follows:
`		ValidationResult result = whenValidating(
				serviceCertificate,
				inter2.getCertificate(),
				inter1.getCertificate(),
				root.getCertificate()
				);
`

Regardless of that issue, the error was in each case the same (if you output validation result, then it is super clear): it was not a problem of chain validation itself, but namespace checking (both Globus and eugridpma). 

So it seems that the whole problem is that openssl validator of CANL is verifying whether all certs from the validated chain are OK wrt namespace policy files. In your testcase it was:
`		given(anOpensslTrustStore()
				.withNamespacesFiles()
				.withSigningPolicyFiles()
				.trustingCA(root)
				.authorising("/DC=org/DC=example")
				//.authorising("/DC=ch/DC=cern")
				.andTrustingCA(inter1)
				.authorising("/DC=org/DC=example"));
`
If I uncomment (added by me) line then everything passes.

To sum up:

1. I can change the code performing namespace validation to check not all the requested certs, but only the chain that was built and actually successfully verified. Such fix is trivial. Personally I have no problem with such change, but I would appreciate your and/or some other grid experts ack here. 
2. Do you think that this can be the real root cause triggering this ticket? I.e. the intermediary cert was causing problems as it was not accepted by your namespaces?

**Comment by msalle** [comment]
Hi @golbi @paulmillar 
I'd say all the certs that are part of the actual chain being used should be validated against all the namespaces that belong to any of the CAs (sub or root) in that specific chain. Any cert and namespace that's not part of that chain can be ignored.

**Comment by paulmillar** [comment]
Hi @golbi ,

Thanks for investigating.

You're right, the certificate order is wrong in the unit-test.  #119  should fix that.

On your two points.

> I can change the code performing namespace validation to not to check all the requested certs, but only the chain that was built and actually successfully verified. 

Excellent.  That is the fix I was hoping for.

> Such fix is trivial. Personally I have no problem with such change, but I would appreciate your and/or some other grid experts ack here.

I think @msalle is happy with this change.  I'm not sure who else we should consult.

> Do you think that this can be the real root cause triggering this ticket? I.e. the intermediary cert was causing problems as it was not accepted by your namespaces?

Yes, I believe so.

This CaNL ticket is specifically about namespace checks on "unnecessary" certificates causing problems.  For me, the underlying issue is the one [reported against dCache](https://github.com/dCache/dcache/issues/6662).

If you're concerned that this issue isn't tackling the underlying/real problem, please feel free to look at the [corresponding dCache issue](https://github.com/dCache/dcache/issues/6662) for all the details.

Note that the actual problem (involving dCache) is messier that the unit-test.  In the interests of making things easy to read/understand, I've tried to simplify the problem: creating a minimal example to demonstrate the problem.  So, it's possible that this unit-test is somehow wrong (doesn't match the problem).

However, I think the problem described in this ticket is real, and (I believe) fixing this issue will go some way towards fixing the dCache issue.

**Comment by msalle** [comment]
Might be good to ask the question on the igtf-general mailing list (ping me offline for the address if you need it).
One reason for including multiple CAs as a server could be to support clients with different CA-distributions (e.g. CAB or IGTF). But that only strengthens the argument that checks should only be done against the chain that actually verifies. For that chain, if it is IGTF, I'd say it's mandatory that namespace verification is still done, since that's an IGTF policy.

**Comment by paulmillar** [comment]
At the risk of redundancy, I've copied a summary of the actual certificate chain the remote server presented (`s` == Subject, `i` == issuer, numbers are the certificate's position in the chain).

```
 0 s:DC = ch, DC = cern, OU = computers, CN = atlas-amazon-cloud.cern.ch
   i:DC = ch, DC = cern, CN = CERN Grid Certification Authority

 1 s:DC = ch, DC = cern, CN = CERN Root CA
   i:DC = ch, DC = cern, CN = CERN Root CA

 2 s:C = ch, O = CERN, CN = CERN Root Certification Authority 2
   i:C = ch, O = CERN, CN = CERN Root Certification Authority 2

 3 s:DC = ch, DC = cern, CN = CERN Grid Certification Authority
   i:C = ch, O = CERN, CN = CERN Root Certification Authority 2

 4 s:DC = ch, DC = cern, CN = CERN Certification Authority
   i:C = ch, O = CERN, CN = CERN Root Certification Authority 2

 5 s:DC = ch, DC = cern, CN = CERN Certification Authority
   i:C = ch, O = CERN, CN = CERN Root Certification Authority 2

 6 s:O = IPAdev CERN, CN = Certificate Authority
   i:O = IPAdev CERN, CN = Certificate Authority
```

This is basically a train-wreck.

On the positive side, the EEC is the first certificate.  However, the certificate order is otherwise wrong (intermediate CA appears after the root CA).  The chain includes root CAs and intermediate CAs that have no bearing on the EEC.  An (unrelated) intermediate CA even appears twice!

Please read my reply in the dCache issue for a fuller analysis, although I now realise I missed the problem with wrong order (oops).

One could (and I did) make the argument that the person running the service should go an fix their certificates.  It's not surprising that something doesn't work.  That was the first (of three) possible solutions presented in the dCache issue.

However, there is still this specific issue: that CaNL-Java rejects the certificate chain because one of the intermediate CAs has the "wrong" DN (per the signing AuthZ files).  I believe C-based clients (for whatever reason) don't suffer from this problem, which is why it was seen as a dCache-specific problem.  However, when I quickly scanned the dCache issue I didn't find a reference to this detail.

**Comment by golbi** [comment]
OK, so the fix was pushed, the test is passing now. @paulmillar do you want to verify whether the lib is now handling your problematic case properly? Let me know, if yes I will wait for your verification, otherwise will proceed with a release.

**Comment by paulmillar** [comment]
Unfortunately, I don't have an easy way to reproduce the problem "in the wild".

I suggest we go with a new CaNL release, which we can then include in dCache.

**Comment by golbi** [comment]
OK, closing then
